File: src/01_data_structures/06_trees/03_trie/implementations/rs/trie.rs
use crate::hash_table::hash_table::HashTable;

#[derive(Debug, PartialEq)]
pub struct Node {
  pub children: HashTable<Node>,
  pub end: bool,
}

impl Node {
  pub fn new() -> Node {
    return Node {
      children: HashTable::new(1),
      end: false,
    };
  }
}

#[derive(Debug, PartialEq)]
pub struct Trie {
  pub root: Box<Node>,
}

impl Trie {
  pub fn new() -> Trie {
    return Trie { root: Box::new(Node::new()) };
  }

  pub fn insert(&mut self, word: &str) -> &mut Self {
    let mut node = &mut self.root;

    for unicode in word.chars() {
      let key = unicode.to_string();
      if !node.children.has(&key) {
        node.children.set(&key, Node::new());
      }
      node = node.children.get_mut(&key).unwrap();
    }

    node.end = true;

    self
  }
  pub fn search(&mut self, word: &str) -> bool {
    let mut node = &mut self.root;

    for unicode in word.chars() {
      let key = unicode.to_string();

      if !node.children.has(&key) {
        return false;
      }
      node = node.children.get_mut(&key).unwrap();
    }

    node.end
  }

  pub fn delete(&mut self, word: &str) -> &mut Self {
    Self::delete_node(&mut self.root, word, 0);

    self
  }

  fn delete_node(node: &mut Node, word: &str, depth: usize) -> bool {
    if word.len() == depth {
      if node.end {
        node.end = false;

        return node.children.is_empty();
      }

      return false;
    }

    let key = word.chars().nth(depth).unwrap().to_string();
    if let Some(child) = node.children.get_mut(&key) {
      let should_delete = Self::delete_node(child, word, depth + 1);
      if should_delete {
        node.children.remove(&key);

        return node.children.is_empty() && !node.end;
      }
    }

    false
  }

  pub fn suggest(&self, prefix: &str) -> Vec<String> {
    let mut node: &Node = self.root.as_ref();

    for unicode in prefix.chars() {
      let key = unicode.to_string();

      if !node.children.has(&key) {
        return vec![];
      }

      node = node.children.get(&key).unwrap();
    }

    let mut results: Vec<String> = Vec::new();
    let mut buffer: Vec<char> = prefix.chars().collect();

    Self::depth_first_breadth(node, &mut buffer, &mut results);

    results
  }

  fn depth_first_breadth(node: &Node, buffer: &mut Vec<char>, results: &mut Vec<String>) {
    if node.end {
      results.push(buffer.iter().collect());
    }

    for bucket in node.children.buckets() {
      for item in bucket.iter() {
        let key_char = item.key.chars().next().unwrap();
        buffer.push(key_char);
        Self::depth_first_breadth(&item.value, buffer, results);
        buffer.pop();
      }
    }
  }
}

File: src/01_data_structures/06_tree/03_trie/implementations/ts/trie.ts
class Node {
  end: boolean = false;
  children: Record<string, Node> = {};
}

export class Trie {
  root: Node = new Node();

  insert(word: string): Trie {
    let node = this.root;

    for (const char of word) {
      if (!node.children[char]) {
        node.children[char] = new Node();
      }

      node = node.children[char];
    }
    node.end = true;

    return this;
  }

  search(word: string): boolean {
    let node = this.root;

    for (const char of word) {
      if (!node.children[char]) {
        return false;
      }

      node = node.children[char];
    }

    return !!node.end;
  }

  delete(word: string): Trie {
    this.deleteNode(this.root, word, 0);

    return this;
  }

  private deleteNode(node: Node, word: string, depth: number): boolean {
    // Reached final node of word
    if (depth === word.length) {
      // If it is marked as end of word
      if (node.end) {
        node.end = false; // Unset `end`, node can be removed.
        // If node has no children, it is leaf, remove it.
        return Object.keys(node.children).length === 0;
      }
    }

    const char = word[depth];
    const child = node.children[char];
    if (!child) return false; // If no child, word not present, nothing to delete.

    const shouldDeleteChild = this.deleteNode(child, word, depth + 1); // Check recursively if next node should be deleted.

    // Delete
    if (shouldDeleteChild) {
      delete node.children[char];
      // If node is leaf, and is not an end of another word, delete it.
      return Object.keys(node.children).length === 0 && !node.end;
    }

    // Otherwise, delete it
    return false;
  }

  suggest(prefix: string): Array<string> {
    const node = prefix
      .split("")
      .reduce((acc, char) => acc.children[char], this.root);
    if (!node) return [];
    const results = [];
    this.depthFirstSearch(node, prefix, results);

    return results;
  }

  depthFirstSearch(node: Node, prefix: string, results: Array<string>) {
    if (node.end) {
      results.push(prefix);
    }

    for (const [char, child] of Object.entries(node.children)) {
      this.depthFirstSearch(child, prefix + char, results);
    }
  }
}
